event-store/src/adapters/pg/store.rs

```diff
@@ -19,133 +21,150 @@ use Events;
-        since: Option<CacheResult<T>>,
-    ) -> (T, String)
-    where
-        E: Events,
-        T: Aggregator<E, A, PgQuery<'a>> + Default,
-        A: Clone,
-    {
-        let (initial_state, query_string) = if let Some((existing, time)) = since {
-            (existing, format!(
-                "SELECT * FROM ({}) AS events WHERE events.context->>'time' >= '{}' ORDER BY events.context->>'time' ASC",
-                query_string.query, time
-            ))
+    fn generate_query(initial_query: &PgQuery<'a>, since: Option<DateTime<Utc>>) -> String {
+        if let Some(timestamp) = since {
+            String::from(format!(
+            "SELECT * FROM ({}) AS events WHERE events.context->>'time' >= '{}' ORDER BY events.context->>'time' ASC",
+            initial_query.query, timestamp,
+        ))
         } else {
-            (T::default(), String::from(query_string.query))
-        };
-
-        (initial_state, query_string)
+            String::from(format!(
+                "SELECT * FROM ({}) AS events ORDER BY events.context->>'time' ASC",
+                initial_query.query
+            ))
+        }
     }
 }
 
 impl<'a> StoreAdapter<PgQuery<'a>> for PgStoreAdapter {
-    fn aggregate<E, T, A>(&self, query_args: A, since: Option<CacheResult<T>>) -> Result<T, String>
+    fn read<'b, E>(
+        &self,
+        query: PgQuery<'b>,
+        since: Option<DateTime<Utc>>,
+    ) -> BoxedFuture<'b, Vec<E>, String>
     where
-        E: Events,
-        T: Aggregator<E, A, PgQuery<'a>> + Default,
-        A: Clone,
+        E: Events + Send + 'b,
     {
-        let q = T::query(query_args);
-        let (initial_state, query_string) = Self::generate_query(&q, since);
-
-        let conn = self.conn.get().expect("Could not get PG connection");
-
-        let trans = conn.transaction().expect("Tranny");
-        let stmt = trans.prepare(&query_string).expect("Prep");
-
-        let mut params: Vec<&ToSql> = Vec::new();
-
-        for (i, _arg) in q.args.iter().enumerate() {
-            params.push(&*q.args[i]);
-        }
-
-        let results = stmt
-            .lazy_query(&trans, &params, 1000)
-            .unwrap()
-            .map(|row| {
-                let id: Uuid = row.get("id");
-                let data_json: JsonValue = row.get("data");
-                let context_json: JsonValue = row.get("context");
-
-                let thing = json!({
-                    "id": id,
-                    "data": data_json,
-                    "context": context_json,
-                });
-
-                let evt: E = from_value(thing).expect("Could not decode row");
-
-                evt
-            }).fold(initial_state, |acc, event| T::apply_event(acc, &event))
-            .expect("Fold");
-
-        trans.finish().expect("Tranny finished");
-
-        Ok(results)
+        let conn = self.pool.get();
+        Box::from(FutLazy(move || {
+            let pool = conn.expect("Could not connect to the pool (aggregate)");
```

Comment:
I think you should get a connection inside the closure, not outside, the reason being that it will keep a slot in the pool reserved for something that may not be executed until later on. To avoid having problems with the lifetimes of self you can clone the pool and use this one inside the closure.
```rust
let pool = self.pool.clone();
FutLazy(move || {
  let conn = pool.get();
})
```